### DISCUSSION TITLE
Router does not work in Micro

### DISCUSSION THREAD

**Discussion by Code8525** [phalcon/cphalcon#15818] [Q&A]
Hello

```
    class ApiController extends Controller
    {
        public function indexAction()
        {
            echo "ApiController indexAction";
        }

    }

    $router = new \Phalcon\Mvc\Router();
    $router->addGet(
        '/api', [
            'controller' => 'api',
            'action' => 'index'
        ]
    );


    $di = new \Phalcon\Di();
    $app = new \Phalcon\Mvc\Micro($di);
    $app->setService('router', $router, true);
    $di->set('request', 'Phalcon\Http\Request');

    $app->handle(
        $_SERVER["REQUEST_URI"]
    );


```

http://site/api -> Matched route doesn't have an associated handler

**Comment by yohowie** [comment]
I also encountered the same problem.
When I checked the source code of cphalcon, I found that the routing of Micro and MVC models are different.

https://github.com/phalcon/cphalcon/blob/master/phalcon/Mvc/Micro.zep#L252
https://github.com/phalcon/cphalcon/blob/master/phalcon/Mvc/Micro.zep#L452

When using mvc routing mode, there is no [handler] parameter in the code.
I don't know if it is the real reason, but the document says it can be used.

**Comment by DavisDS** [comment]
I had the same issue, but I found some old lib written to support annotations and reverse engineered the hell out of it.  Short version is the way micro mounts routes is totally different than the router object.  Even if you scale a medium application on Micro it's very different than a full application.  

If you are interested in using annotations, I have a production release version ready that allows annotation routes with Micro.  It supports caching too.

[phalcon-micro-annotations](https://github.com/SageITSolutions/phalcon-micro-annotations)

**Comment by secgin** [comment]
Hello, I’ve encountered this issue as well. When I checked the code in the Phalcon GitHub repository, I noticed that the Micro class clears the routes before returning the router service through the getRouter method. I couldn’t understand why this was done.

Secondly, in the handle method of the Micro class, it checks if the route exists within this->handlers. When you add routes to the Micro class, they are stored in the this->handlers array.

The documentation mentions that we should create and include the Router class, so I wanted to share this information.

Best regards and good luck with your work!
